cudaaligner/src/myers_gpu.cu

```diff
@@ -0,0 +1,433 @@
+/*
+* Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include "myers_gpu.cuh"
+#include <cassert>
+#include <climits>
+#include <vector>
+#include <numeric>
+#include <utils/signed_integer_utils.hpp>
+#include "device_storage.cuh"
+#include "batched_device_matrices.cuh"
+#include <cudautils/cudautils.hpp>
+
+namespace claragenomics
+{
+
+namespace cudaaligner
+{
+
+namespace myers
+{
+
+inline __device__ WordType warp_leftshift_sync(uint32_t warp_mask, WordType v)
+{
+    constexpr int32_t word_size = sizeof(WordType) * CHAR_BIT;
+    const WordType x            = __shfl_up_sync(warp_mask, v >> (word_size - 1), 1);
+    v <<= 1;
+    if (threadIdx.x != 0)
+        v |= x;
+    return v;
+}
+
+inline __device__ WordType warp_add_sync(uint32_t warp_mask, WordType a, WordType b)
+{
+    static_assert(sizeof(WordType) == 4);
+    static_assert(CHAR_BIT == 8);
+    const uint64_t ax = a;
+    const uint64_t bx = b;
+    uint64_t r        = ax + bx;
+    uint32_t carry    = static_cast<uint32_t>(r >> 32);
```

Comment:
should this `32` (and another carry command one line 57) be replaced with something that's related to the wordsize?

Reply:
I think this is fine. The static_assert (line 42) states that the function expects WordType to be 4 bytes and the variable names suggest the intent.

Reply:
okie dokes!